src/common/snap_types.cc

```diff
@@ -12,6 +12,8 @@ void SnapRealmInfo::encode(ceph::buffer::list& bl) const
   encode(h, bl);
   ceph::encode_nohead(my_snaps, bl);
   ceph::encode_nohead(prior_parent_snaps, bl);
+  encode(last_modified, bl);
+  encode(change_attr, bl);
```

Comment:
Maybe we should add one feature bit and only when the client supported it will encode these two members ?

Reply:
I have the changes (feature bit) locally, but I'm not sure why tests were failing even when this change was dropped from the branch (I had removed the merge commit before building). Building a fresh branch now...

Reply:
Oh, and BTW, I will push out the feature bit changes before putting this to test.

Reply:
... and some kclient tests ran to completion. I need to go back and check those.

Reply:
Yeah, feature bit is a must for now.

I suggest to add the version and compat version for `SnapRealmInfo::encode()/decode()`. Then we can just bump up the version instead of adding new feature bits again and again in future if we add or change the members.

Reply:
We discussed that here - https://github.com/ceph/ceph/pull/48382#discussion_r1006412731

Reply:
BTW, the way snap traces are maintained are that the encoded snap traces are kept in a cache, so, that makes it a bit tricky to selectively encode members in a snap realm based on client features.

Reply:
> We discussed that here - [#48382 (comment)](https://github.com/ceph/ceph/pull/48382#discussion_r1006412731)

Okay, I forgot that.

> BTW, the way snap traces are maintained are that the encoded snap traces are kept in a cache, so, that makes it a bit tricky to selectively encode members in a snap realm based on client features.

Yeah. How about adding a new cached extra snap traces ?

And just add a new versioned extra struct:

```
mutable ceph_mds_snap_realm_extra {
  utime_t last_modified;
  uint64_t change_attr;
};
```
And when encoding this we should version it. And then when replying to client request just do:

```
reply->snapbl = realm->get_snap_trace();
if (client support extra snap trace) {
    reply->snapbl.append(realm->get_snap_trace_extra())
}
```
??

Reply:
Yeh, something like that or introduce a new bufferlist in MClientReply and leave reply->snapbl untouched.

Reply:
Or we need to build the snap trace cache again and again if there have old clients and new clients connected.

Reply:
> Or we need to build the snap trace cache again and again if there have old clients and new clients connected.

That would not be ideal IMO. Mixed clients (which is _very very_ likely) would unnecessarily result in cache being regenerated.

Reply:
How about maintaining a "versioned" SnapRealmInfo along side current SnapRealmInfo? Then the MDS could send either the versioned struct or otherwise?

Reply:
Yeah, this also sounds good to me.